File: app/src/am_hieunt/com/example/asian/issue8/ViewModel/UserViewModel.java
package com.example.asian.issue8.ViewModel;

import android.annotation.SuppressLint;
import android.content.Context;

import com.example.asian.issue8.Model.User;
import com.example.asian.issue8.Model.UserDatabase;

import java.util.List;

public class UserViewModel {
    @SuppressLint("StaticFieldLeak")
    private static UserViewModel instance;
    private final Context context;

    public static UserViewModel getInstance(Context context) {
        if (instance == null) {
            instance = new UserViewModel(context);
        }
        return instance;
    }

    private UserViewModel(Context context) {
        this.context = context;
    }

    public User getUser(int id) {
        return UserDatabase.getInstance(context).getUser(id);
    }

    public List<User> getAllUser() {
        return UserDatabase.getInstance(context).getAllUser();
    }

    public void insertUser(User user) {
        UserDatabase.getInstance(context).insertUser(user);
    }

    public void updateUser(User user) {
        UserDatabase.getInstance(context).updateUser(user);
    }

    public void deleteUser(int id) {
        UserDatabase.getInstance(context).deleteUser(id);
    }

    public void deleteAllUser() {
        UserDatabase.getInstance(context).deleteAllUser();
    }
}

File: app/src/am_hieunt/com/example/asian/issue13/ex1/viewmodel/UserViewModel.kt
package com.example.asian.issue13.ex1.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.viewModelScope
import com.example.asian.issue13.ex1.database.UserDatabase
import com.example.asian.issue13.ex1.model.User
import com.example.asian.issue13.ex1.repository.UserRepository
import kotlinx.coroutines.launch

class UserViewModel(application: Application): AndroidViewModel(application) {
    val allUser: LiveData<List<User>>
    private val repository: UserRepository
    init {
        val userDAO = UserDatabase.getDatabase(application).userDAO()
        repository = UserRepository(userDAO)
        allUser = repository.allUser
    }
    fun addUser(user: User) = viewModelScope.launch {
        repository.insertUser(user)
    }
    fun deleteUser(user: User) = viewModelScope.launch {
        repository.deleteUser(user)
    }
    fun deleteAllUser() = viewModelScope.launch {
        repository.deleteAllUser()
    }
    fun updatePosition(users: List<User>, userDeleteId: Int): List<User> {
        var index = 0
        val newUsers = mutableListOf<User>()
        for (user in users) {
            if (user.id > userDeleteId) {
                val us = User(user.name, user.age)
                us.id = user.id
                us.position = index
                newUsers.add(us)
            } else {
                newUsers.add(user)
            }
            index++
        }
        return newUsers
    }
}
